pyvista/core/utilities/geometric_objects.py

```diff
@@ -363,6 +375,463 @@ def Sphere(
+    `Sphere Point Picking <https://mathworld.wolfram.com/SpherePointPicking.html>`_.
+
+    >>> phi = np.rad2deg(np.arccos(np.linspace(1, -1, 30)))
+    >>> solid_sphere = pv.SolidSphereGeneric(
+    ...     radius=np.linspace(0, 0.5, 2),
+    ...     theta=np.linspace(180, 360, 30),
+    ...     phi=phi,
+    ... )
+    >>> solid_sphere = solid_sphere.compute_cell_sizes()
+    >>> solid_sphere.plot(
+    ...     scalars="Volume", show_edges=True, clim=[3e-5, 5e-4]
+    ... )
+
+    """
+    radius = np.asanyarray(radius)
+    theta = np.asanyarray(theta)
+    phi = np.asanyarray(phi)
+
+    nr = len(radius)
+    ntheta = len(theta)
+    nphi = len(phi)
+
+    if nr < 2:
+        raise ValueError("radius resolution must be 2 or more")
+    if ntheta < 2:
+        raise ValueError("theta resolution must be 2 or more")
+    if nphi < 2:
+        raise ValueError("phi resolution must be 2 or more")
+
+    is_sorted = lambda a: np.all(a[:-1] < a[1:])
+    if not is_sorted(radius):
+        raise ValueError("radius is not monotonically increasing")
+    if not is_sorted(theta):
+        raise ValueError("theta is not monotonically increasing")
+    if not is_sorted(phi):
+        raise ValueError("phi is not monotonically increasing")
+
+    if radius[0] < 0.0:
+        raise ValueError("minimum radius cannot be negative")
+    if theta[0] < 0.0:
+        raise ValueError("minimum theta cannot be negative")
+    if phi[0] < 0.0:
+        raise ValueError("minimum phi cannot be negative")
+    if theta[-1] > 360.0:
+        raise ValueError("maximum theta cannot be > 360.0")
+    if phi[-1] > 180.0:
+        raise ValueError("maximum phi cannot be > 180.0")
+
+    # Hereafter all degrees are in radians
+    theta = np.deg2rad(theta)
+    phi = np.deg2rad(phi)
+
+    def _spherical_to_cartesian(r, phi, theta):
+        """Convert spherical coordinate sequences to a ``(n,3)`` cartesian coordinate array.
+
+        Parameters
+        ----------
+        r : sequence[float]
+            Ordered sequence of floats of radii.
+        phi : sequence[float]
+            Ordered sequence of floats for phi direction.
+        theta : sequence[float]
+            Ordered sequence of floats for theta direction.
+
+        Returns
+        -------
+        np.ndarray
+            ``(n, 3)`` cartesian coordinate array.
+
+        """
+        r, phi, theta = np.meshgrid(r, phi, theta, indexing='ij')
+        x = r * np.sin(phi) * np.cos(theta)
+        y = r * np.sin(phi) * np.sin(theta)
+        z = r * np.cos(phi)
+        return np.vstack((x.ravel(), y.ravel(), z.ravel())).transpose()
+
+    points = []
+
+    npoints_on_axis = 0
+
+    if radius[0] == 0.0:
+        points.append([0.0, 0.0, 0.0])
+        include_origin = True
+        nr = nr - 1
+        radius = radius[1:]
+        npoints_on_axis += 1
+    else:
+        include_origin = False
+
+    if theta[0] == 0.0 and theta[-1] == 2 * np.pi:
```

Comment:
My plan is to look into how the merge points tolerance works and then make it user configurable.

Reply:
Yeah, it seems like the tolerance needs to be user configurable otherwise the tolerance might exclude some case the user is actually trying to do. But then is a configurable tolerance needed at r=0, phi=0, phi=pi and theta[0] = theta[-1]? Maybe the tolerance parameter can accept either a float (which will be used for all 4) or a list of 4 floats which will be applied separately to each. Also, does the user need to be able to specify `atol` and `rtol` if `np.isclose()` is used?

Kind of super annoying. Maybe you could use `np.isclose()` for now and add a feature for user configurable tolerances in a future PR where there could be a separate discussion?

Reply:
> But then is a configurable tolerance needed at r=0, phi=0, phi=pi and theta[0] = theta[-1]?

Not for zeros I think, but perhaps for nonzeros. `theta[-1]` can only equal `theta[0]` mod 2 pi...



> Also, does the user need to be able to specify `atol` and `rtol` if `np.isclose()` is used?

Hopefully not, because the exact API of `np.isclose()` is a bit weird.

Reply:
We want the function to 'just work' in the majority cases, but also to not constrain users with semi-arbitrary limits.  For example, what if the user truly wants the SolidSphere to be incomplete in the `theta` direction, but only by a small amount?  We also want to have a check that works at both `0` and nonzero, like `pi` or `2 * pi`.  If we want to allow theta outside `[0, 2*pi]`, it will require a close-like check.  My understanding from reading the vtk delegated operations like merge and clean is that they use an absolute tolerance only for this kind of check.  For example, this method below uses absolute tolerance or uses a relative tolerance, but it is relative to the whole dataset size and is not used in the 'relative tolerance' sense that `np.isclose` uses it.

https://vtk.org/doc/nightly/html/classvtkStaticCleanUnstructuredGrid.html#a8fdb401493b5fd01dc48ddfadeb75cbe


Since the numbers are all order ~1 or less, between 0 and 2*pi to be precise, **I propose a configurable absolute tolerance of 1e-8, i.e the default atol of np.isclose**.  This probably just works for 99.9% of cases.  In the other 0.1% of cases, the user can tune the value or set it to 0.  We will end up with a check like `theta[0] % (2 * pi)` == `theta[-1] % (2 * pi)` which will have numerical issues with this definition if someone defines theta to be order ~1e9.  We can caution that unexpected results might be obtained with very large theta due to endpoint handling.

`vtkSphereSource` interestingly does a greater than or equal to check for the upper bound.  For example, if `phi >= pi`, they consider it on the axis.  I'm not proposing this as a solution, but wanted to give a full picture here.

https://github.com/Kitware/VTK/blob/b343a749a122f02e5f98a9d0020b33ea2e18a968/Filters/Sources/vtkSphereSource.cxx#L155-L156

Reply:
I implemented this.  One subtlety in the implementation is that it uses the tolerance parameter to detect overlap of points, but it doesn't 'snap' `phi` to `0` or `2*pi`.  It keeps the precision of phi provided.  So for example, a value of `start_phi=1e-9` will result in an axis but it will be slightly tilted at `phi=1e-9` in the positive z-direction and `phi=np.pi` in the negative z-direction.  Note that these axes aren't colinear.

Reply:
> `vtkSphereSource` interestingly does a greater than or equal to check for the upper bound. For example, if `phi >= pi`, they consider it on the axis.

I suspect this is just a two-for-one fix for numerical inaccuracies. E.g. I've managed to get errors from `np.arccos()` when the dot product of two unit vectors was slightly lower than -1 (i.e. slightly larger than 1 in absolute value) due to floating-point errors. The fix in that case would be to `np.clip()` the argument to `[-1, 1]`, and VTK could've done the same with the angle, but it probably was enough for them to catch everything over `pi`.